ServerSideApply E2E tests

Jefftree opened the pull request:
**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:

SSA E2E tests

Test list:
- Apply create
- Apply subresource
- Removing ownership of field single owner
- Removing ownership of field multiple owners
- Conflict & force apply
- Interaction between apply/update (for controllers)
- Apply CRD (multi version + atomic support)

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

Jefftree committed:
Add E2E tests for SSA

Jefftree commented:
/cc @apelisse

apelisse commented:
I'm not an expert in e2e tests nor conformance test but that looks reasonable to me.

cc @liggitt

fedebongio commented:
/triage accepted

liggitt commented on test/e2e/apimachinery/apply.go:
```diff
@@ -0,0 +1,302 @@
+			Name("deployment").
+			Param("fieldManager", "apply_test").
+			Body([]byte(obj)).Do(context.TODO()).Get()
+		if err == nil {
```
make sure the error is a conflict

Jefftree replied:
Updated

liggitt reviewed (commented):
looked through quickly, just had one comment about checking the type of error is actually a conflict

Jefftree commented:
Addressed the comments. I think it'd be useful to have this merged to serve as a good example for other SSA tests.

/assign @liggitt

liggitt commented on test/e2e/apimachinery/apply.go:
```diff
@@ -0,0 +1,303 @@
+/*
+Copyright 2017 The Kubernetes Authors.
```
2021

liggitt commented on test/e2e/apimachinery/apply.go:
```diff
@@ -0,0 +1,303 @@
+					framework.Failf("Failed to retrieve object: %v", err)
+				}
+
+				// Test that we can re apply with a different field manager and don't get conflicts
+				_, err = client.CoreV1().RESTClient().Patch(types.ApplyPatchType).
```
should this be inspecting the result for some expected state? do we expect two field managers that co-own these fields, or that apply_test_2 owns no fields (since it wasn't changing any state)?

Jefftree replied:
cc @apelisse, I think the two managers will co own the fields?

From a conformance perspective, managed fields are opaque and shouldn't be something users need to directly deal with. Wouldn't it make more sense to try to perform operations after a specific state is reached instead of directly inspecting the managed fields? (Eg: What happens when one of the managers applies something afterwards?) I added another test for the case when two managers own a field. When one manager drops ownership of that field, the field should not be deleted.

apelisse replied:
> cc @apelisse, I think the two managers will co own the fields?

Yes.

> From a conformance perspective, managed fields are opaque

They are part of the API at this point, they can't be treated as opaque.

Jefftree replied:
ack, added a check for the managedFields. It compares the stringified managed fields with a const string. Not sure if that's the ideal way to do it?

liggitt commented on test/e2e/apimachinery/apply.go:
```diff
@@ -0,0 +1,303 @@
+		}
+	})
+
+	ginkgo.It("PUT requests which change nothing should not change the resource version", func() {
```
we have TestNoOpUpdateSameResourceVersion as an integration test... do we need an e2e as well?

Jefftree replied:
Yeah I agree the two tests should either both be in e2e or both in integration, it doesn't make sense to move just one of them. Moved this one outside this PR for now, can add it back in later if we need.

liggitt commented on test/e2e/apimachinery/apply.go:
```diff
@@ -0,0 +1,303 @@
+			framework.Failf("Expecting to get conflicts when applying object")
+		}
+		status, ok := err.(*apierrors.StatusError)
+		if !(ok && status.ErrStatus.Reason == metav1.StatusReasonConflict) {
+			framework.Failf("Expecting to get conflicts as API error")
+		}
+		if len(status.Status().Details.Causes) < 1 {
+			framework.Failf("Expecting to get at least one conflict when applying object, got: %v", status.Status().Details.Causes)
+		}
```
apierrors.IsConflict will do this check

liggitt commented on test/e2e/apimachinery/apply.go:
```diff
@@ -0,0 +1,303 @@
+		}
+	})
+
+	ginkgo.It("Force apply should ignore API errors ", func() {
```
this is specifically ignoring conflict errors, right? should probably use that as the test title, not API errors generally

liggitt reviewed (commented):
what determines which tests we'll leave as lighter-weight (and easier to run) integration tests, and which will be duplicated/converted to e2e tests?

Jefftree commented:
My take is that the e2e tests should cover the basic list of expected behaviors for user visible operations. Specific edge cases and things around performance can remain in integration tests.

Here's the list I'm thinking:
- Normal Apply (pre-existing and non-existent object)
- Apply Conflict (and --force option)
- Apply Subresource
- Integration with controllers
- Normal Apply on CRDs

jpbetz commented on test/e2e/apimachinery/apply.go:
```diff
@@ -0,0 +1,303 @@
+	ginkgo.AfterEach(func() {
+	})
+
+	ginkgo.It("PATCH requests with the apply content type will create the object if it does not already exist", func() {
```
We can start getting the test identifiers and conformance names and descriptions that we will need for conformance ready now too if we have time.

Descriptions should use rfc2119 language (MUST, MAY, SHOULD) I believe. More details at https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md#conformance-test-comment-metadata. I don't see much guidance on how to format testname..

For test identifiers `ginkgo.It(` in intended to be interpreted as starting a sentence with "It..", e.g. "It should be able to create an object using an patch request if it doesn't exist already", where "it" typically refers to whatever was passed to `SIGDescribe()`.
 
Some examples: https://github.com/kubernetes/kubernetes/pull/81864/files

Jefftree replied:
Thanks for the suggestion and links. I'll try to include the conformance block in this PR as well.

jpbetz commented:
> Test list:
> 
> * Apply create
> * Apply subresource
> * Removing ownership of field single owner
> * Removing ownership of field multiple owners
> * Conflict & force apply
> * Interaction with apply/update (for controllers)
> * Apply CRD

This is a good list. The only other things I could think of are:

- multi-version support (In conformance we can only use stable APIs, probably needs to be tested with CRDs)
- atomic vs. granular (e.g. make sure something like LabelSelectors has atomic apply semantics)
- merging rules (two appliers can both edit their own items in an associative list without conflicting)
- field ownership transfers (https://kubernetes.io/docs/reference/using-api/server-side-apply/#transferring-ownership)

Jefftree commented:
Thanks for the feedback Joe! I added tests to address your comments. 

> * multi-version support (In conformance we can only use stable APIs, probably needs to be tested with CRDs)

Added within the CRD test

> * atomic vs. granular (e.g. make sure something like LabelSelectors has atomic apply semantics)

The GA atomic lists I found in k/k were all immutable. I created a CRD to test this instead

> * merging rules (two appliers can both edit their own items in an associative list without conflicting)

This is somewhat covered by the subresource test on status.conditions. I'm not sure if it's worth writing another test?

Jefftree commented:
/retest

jpbetz commented on test/e2e/apimachinery/apply.go:
```diff
@@ -0,0 +1,1181 @@
+	/*
+		Release : v1.21
+		Testname: Server Side Apply, unset field shared
+		Description: Apply an object. Unset ownership of a field that is also owned by other managers and make a subsequent apply request. The unset field MUST not be removed from the object.
```
We could verify the managed fields are what we expect as well to ensure the ownership transfer really worked.

Jefftree replied:
It's fairly verbose the check the managedFields without importing SMD. I think it should be enough that removing the field from the first applier didn't delete the field to recognize that the field is still owned by the other applier.

jpbetz replied:
I agree. Thanks for checking.

jpbetz commented:
Looks great. Coverage of SSA features is good. Tests are well named and documented.
Just one minor suggestion about leveraging managed fields to check ownership.

jpbetz commented:
/lgtm

lavalamp commented:
I skimmed, looks OK. I think I would prefer it if the tests were more dynamic--e.g. right now you could pass them more or less by replaying a sequence of responses--but I don't think that's super important right now.

/approve
